### PR TITLE
Added ability for ZoneEntities to change gravity

### PR DESCRIPTION
**Pull request by hyperlogic**
This PR adds a new gravity section to zone entities.  This allows the content creator to modify the direction and magnitude of gravity for all physical objects within the zone.  Example uses:  
  * Reduced gravity in a moon like domain.
  * A spherical planetoid domain were gravity always points toward the center of the planet.
  * An attractor that pulls objects toward a fake magnet or vortex.

This only affects dynamic objects, not avatars or particle systems.

NOTE: this branch is work in progress.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3949/HighFidelity-Beta-PR11541-24fa0c7f43a883df0962351b0b56f3b59bfca54d.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3962/HighFidelity-Beta-PR11541-577d1c681a88bf3291370b2ff732fff6d027cafd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3962/HighFidelity-Beta-PR11541-577d1c681a88bf3291370b2ff732fff6d027cafd.exe

File: libraries/networking/src/udt/PacketHeaders.h [hunk 1]
```diff
@@ -268,6 +268,7 @@ const PacketVersion VERSION_ENTITIES_HAS_SHOULD_HIGHLIGHT = 71;
 const PacketVersion VERSION_ENTITIES_HAS_HIGHLIGHT_SCRIPTING_INTERFACE = 72;
 const PacketVersion VERSION_ENTITIES_ANIMATION_ALLOW_TRANSLATION_PROPERTIES = 73;
 const PacketVersion VERSION_ENTITIES_HAS_CERTIFICATE_PROPERTIES = 74;
+const PacketVersion VERSION_ENTITIES_HAS_ZONE_GRAVITY_PROPERTIES = 75;
```
Comment (AndrewMeadows) [inline libraries/networking/src/udt/PacketHeaders.h]: Wow, most of those code changes above this ^^^ were boiler plate required for adding new properties.

File: libraries/physics/src/EntityMotionState.cpp [hunk 2]
```diff
@@ -46,6 +46,14 @@ bool entityTreeIsLocked() {
 #endif
 
 
+#define CRASH_ASSERT(exp)                      \
+    do {                                       \
+        if (!(exp)) {                          \
+            int* ptr = (int*)0;                \
+            *ptr = 0x0badf00d;                 \
+        }                                      \
+    } while (0)
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.cpp]: `CRASH_ASSERT` is enabled for Release builds.  Is that intentional?

File: libraries/physics/src/GravityZoneAction.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    // these are from btActionInterface
+    virtual void updateAction(btCollisionWorld* collisionWorld, btScalar deltaTimeStep) override;
+    virtual void debugDraw(btIDebugDraw* debugDrawer) override;
+protected:
+
+    std::unique_ptr<btBoxShape> _box;
+    AddRemovePairGhostObject _ghost;
+    btDynamicsWorld* _world { nullptr };
+    ZonePhysicsActionProperties _zpap;
```
Comment (AndrewMeadows) [inline libraries/physics/src/GravityZoneAction.h]: The `_zpap` variable name annoys me, but I'm not sure I have a better name to suggest.  Maybe just `_properties` but I could see how that might be confusing.

Comment (hyperlogic) [reply libraries/physics/src/GravityZoneAction.h]: I can change it to _zoneActionProperties

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3972/HighFidelity-Beta-PR11541-a0f214dfc8e8f2878f60fd9ac46c56f4f65ff3b9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3972/HighFidelity-Beta-PR11541-a0f214dfc8e8f2878f60fd9ac46c56f4f65ff3b9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3979/HighFidelity-Beta-PR11541-7d39cfc3aef2ace48dda0be594257fab538684a5.dmg

**Comment by hyperlogic** [comment]
good ole error 23:

    fatal error C1090: PDB API call failed, error code '23': '( [C:\jenkins\workspace\pr-build\label\windows\build\interface\interface.vcxproj]

lets re-build this.  build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3987/HighFidelity-Beta-PR11541-7d39cfc3aef2ace48dda0be594257fab538684a5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3987/HighFidelity-Beta-PR11541-7d39cfc3aef2ace48dda0be594257fab538684a5.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3989/HighFidelity-Beta-PR11541-06862995920e45c72a63f273352673e38362d4cc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3989/HighFidelity-Beta-PR11541-06862995920e45c72a63f273352673e38362d4cc.exe

**Comment by hyperlogic** [comment]
QA Test:
  * Start with the local sandbox domain.
  * Use "Create" toolbar item to create a zone entity that is 10x10x10.
  * In the "Create" properties tab, set this zone entity to have Linear gravity and a gravity up direction of (0,1,0)
  * Using the marketplace search for "gnome" add the [Box 'o Gnomes](https://metaverse.highfidelity.com/marketplace/items/de5d3039-6495-47b1-bfc5-e2949347f4f5) item to the world.
  * Using the mouse or hand controllers, drag one of the gnomes into the zone.
  * The gnome should rise into the air until it leaves the top of the zone, then it should fall back down.
  * Any objects put into the zone should repeatedly rise and fall
  * Ensure that other users in the domain see the same position/orientation of the gnome as you do.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:
[Used 3 computers total, Aaron x 2, Cain x 1]

Start with the local sandbox domain. **Check**
Use "Create" toolbar item to create a zone entity that is 10x10x10. **Check**
In the "Create" properties tab, set this zone entity to have Linear gravity and a gravity up direction of (0, -1, 0) **Check**
Using the marketplace search for "gnome" add the Box 'o Gnomes item to the world. **Check**
Using the mouse or hand controllers, drag one of the gnomes into the zone. **Check**
The gnome should rise into the air until it leaves the top of the zone, then it should fall back down. **Check**
Any objects put into the zone should repeatedly rise and fall **Check, They don't fall all the way to the ground, they somewhat hover up and down up where the tip of the zone is**
Ensure that other users in the domain see the same position/orientation of the gnome as you do.**Check, checked with 2 machines my end, 1 from Cain**

**Pass**

**Comment by AndrewMeadows** [comment]
This new zone gravity override feature kinda conflicts with the current **gravity** feature.  The gravity of an entity can be set on a per-entity basis, and specifies the actual world-frame gravity vector.  The zone-specific gravity will override this direction.  If we merge this then eventually we'll have to reconcile these two features, else we'll forever have introduced a "gotcha" or "bug" that will bite content creators.  I have an idea on how to do this, but it would require overhauling the EntityItem gravity property which may break content and will require some backward compatibility for loading old EntityItem format: in other words, it won't be much fun to implement.

All that said, I think we can merge this and move forward.  We'll just have to remember to get around to fixing gravity properly.

**Comment by ZappoMan** [comment]
uh --- can we get a meeting about the spec on this? who's the PM that owns this? where's the spec... in particular... to Andrew's point... there are already per-entity "gravity" -- how do these things related?

File: libraries/animation/src/Rig.cpp [hunk 4]
```diff
@@ -301,7 +301,9 @@ void Rig::setModelOffset(const glm::mat4& modelOffsetMat) {
         _rigToGeometryTransform = glm::inverse(_geometryToRigTransform);
 
         // rebuild cached default poses
-        buildAbsoluteRigPoses(_animSkeleton->getRelativeDefaultPoses(), _absoluteDefaultPoses);
+        if (_animSkeleton) {
```
Comment (ZappoMan) [inline libraries/animation/src/Rig.cpp]: this looks like a change that's unrelated to this PR... shouldn't this be in a different PR with a specific test plan for animations?

File: libraries/entities/src/ZonePhysicsActionProperties.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+struct ZonePhysicsActionProperties {
+    enum Type { None = 0, Spherical, Linear };
+
+    struct SphericalGravityZone {
+        float gforce;
+    };
+
+    struct LinearGravityZone {
+        float gforce;
+        float up[3];
```
Comment (ZappoMan) [inline libraries/entities/src/ZonePhysicsActionProperties.h]: why isn't this a glm::vec3?

**Comment by hyperlogic** [comment]
https://highfidelity.fogbugz.com/f/cases/7592/IPD-Tertiary-Task-List-Entity-Gravity-Not-Applied-in-the-Right-Direction

File: libraries/animation/src/Rig.cpp (hunk 4 above)
Comment (hyperlogic) [reply libraries/animation/src/Rig.cpp]: I spoke with Seth specifically about this, and we determined that this change was too small to warrant a separate PR

File: libraries/entities/src/ZonePhysicsActionProperties.h (hunk 5 above)
Comment (hyperlogic) [reply libraries/entities/src/ZonePhysicsActionProperties.h]: Because it is used inside of a union.  You cannot have classes with constructors in side a union.

File: libraries/animation/src/Rig.cpp (hunk 4 above)
Comment (sethalves) [reply libraries/animation/src/Rig.cpp]: this was one of the things that's causing "runtime error" warnings when I run clang-built interface on Linux.  I don't think it's actually hurting anything -- it just reduces my signal-to-noise ratio.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4565/HighFidelity-Beta-PR11541-96ca978b714f5fe020249307c81eb7dbd90085e7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4565/HighFidelity-Beta-PR11541-96ca978b714f5fe020249307c81eb7dbd90085e7.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4577/HighFidelity-Beta-PR11541-186337a5d6a2f236eb26bfa292047b513c10481e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4621/HighFidelity-Beta-PR11541-7d3dfe23303fa86c2cc3b6b5202830230bc88c7f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4682/HighFidelity-Beta-PR11541-c15bb4834471e37f6f3289c6f736533923621333.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4683/HighFidelity-Beta-PR11541-e349fd53ef310726ea44414238469c01488c166c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4683/HighFidelity-Beta-PR11541-e349fd53ef310726ea44414238469c01488c166c.exe

**Comment by ZappoMan** [comment]
removing the target milestone until the spec has been agreed to

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4703/HighFidelity-Beta-PR11541-f5913bad4595b5a6e9c03fa3ff8dc111d6dbec37.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4703/HighFidelity-Beta-PR11541-f5913bad4595b5a6e9c03fa3ff8dc111d6dbec37.exe

File: libraries/physics/src/PhysicalEntitySimulation.cpp [hunk 6]
```diff
@@ -92,29 +111,33 @@ void PhysicalEntitySimulation::changeEntityInternal(EntityItemPointer entity) {
...
-    if (motionState) {
-        if (!entity->shouldBePhysical()) {
-            // the entity should be removed from the physical simulation
-            _pendingChanges.remove(motionState);
-            _physicalObjects.remove(motionState);
-            _outgoingChanges.remove(motionState);
-            _entitiesToRemoveFromPhysics.insert(entity);
-            if (entity->isMovingRelativeToParent()) {
-                _simpleKinematicEntities.insert(entity);
+    if (entity->getType() == EntityTypes::Zone) {
```
Comment (ZappoMan) [inline libraries/physics/src/PhysicalEntitySimulation.cpp]: technically zones can also be parented and be moving (and I guess physical)... does this change cause zones to no longer support those features?

Comment (hyperlogic) [reply libraries/physics/src/PhysicalEntitySimulation.cpp]: Parenting and moving should work, I'll make sure to add this to the test cases.

Comment (hyperlogic) [reply libraries/physics/src/PhysicalEntitySimulation.cpp]: The shape type is missing from edit.js so it's kind of hard to test "physical" zones.  I'll have to switch to master to verify.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4725/HighFidelity-Beta-PR11541-d64351bbe7f58146f8c973016a9b013d306f5dc8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4725/HighFidelity-Beta-PR11541-d64351bbe7f58146f8c973016a9b013d306f5dc8.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4745/HighFidelity-Beta-PR11541-eef7a265d01001da9e347dab7a702eb3d2487611.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4745/HighFidelity-Beta-PR11541-eef7a265d01001da9e347dab7a702eb3d2487611.dmg

File: libraries/animation/src/Rig.cpp (hunk 4 above)
Comment (hyperlogic) [reply libraries/animation/src/Rig.cpp]: this has been moved to #11783

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4810/HighFidelity-Beta-PR11541-fa56eb123c7d529523acac9d72ab3b34831235ac.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4810/HighFidelity-Beta-PR11541-fa56eb123c7d529523acac9d72ab3b34831235ac.exe

**Comment by hyperlogic** [comment]
This project was not approved.  good bye gravity zones. 😢
